### ISSUE TITLE
Refactor how compiler picks visitors

### ISSUE THREAD

**Issue by wooorm** [wooorm/remark#133]
Currently, the compiler has both utilities, and visitors (by their corresponding node’s type) on its prototype. The good thing about that is that it makes it very easy to subclass constructors to compile something other than markdown while still being able to use some of the provided utilities (like `encode`).

However, the bad side is that when nodes have a type set to the name of a utility (for example, `escape`), that function will be invoked by `visit` to compile the node. And I haven’t thought of nodes with a type of `constructor`.

This issue was originally raised in a different repo. Check that out for the back story: wooorm/mdast-util-escape#1.

The current solution is not good enough because:

*   The possible names of custom nodes (e.g., as defined by plug-ins) are limited by properties and methods on the compiler.

The proposed solutions must:

*    Allow nodes with types in `Object.prototype` (`constructor`), the compiler (`compile`), or properties `file`;
*    Allow plug-ins to overwrite visitors (function which handle the compilation of a node);
*    Allow plug-ins to use utilities like `encode` and `escape`;

Supersedes wooorm/mdast-util-escape#1

**Comment by wooorm** [comment]
@eush77, regarding https://github.com/wooorm/mdast-util-escape/issues/1#issuecomment-172228940, I don’t think it would be necessary to copy all compilers over. Just the ones you’re planning on overwriting, because `super` creates a new object when invoked. Did you miss that, or am I now missing your point :wink:

**Comment by eush77** [comment]
@wooorm I guess I just don't understand advantages and disadvantages of initializing `visitors` in constructor vs. ahead-of-time.

In both cases you can add a visitor in constructor:

```js
class CommentBlockCompiler extends Compiler {
  constructor(file, options) {
    super(file, options);

    this.visitors.root = function (node) {
      // ...
    };
  }
}
```

But in the second case you also can do it out of constructor:

```js
class CommentBlockCompiler extends Compiler {
  constructor(file, options) {
    super(file, options);
  }
}

CommentBlockCompiler.prototype.visitors.root = function (node) {
  // ...
};
```

or even:

```js
Compiler.prototype.visitors.root = function (node) {
  // ...
};
```

Can you help me clarify which is better and why?

**Comment by wooorm** [comment]
Ah, I mistakingly based my thinking on the fact that, when the `visitors` object is on the prototype, and when constructing a new instance, that `visitors` object would be the same for multiple instances of `remark`. Thus, when a plug-in mutates it in one processor, it would change for all other processors too. 

This thinking is correct, however, under the hood the prototype is [cloned](https://github.com/wooorm/unherit/blob/03c7c3e7ff84f443f7418e39d28fb45f859d660a/index.js#L38) ([deeply](https://github.com/pvorb/node-clone)), resulting in each processor having a custom `Compiler` which when its prototype is mutated does not result in the `Compiler`s of other processors to be mutated.

Tl;DR, all above methods should work fine without affecting other processors. So we can opt for the best looking one. Which is your current preference?

**Comment by eush77** [comment]
If there are no other reasons to prefer one over another, I would like if the process of adding a visitor to Compiler mirrored the process of [adding a tokenizer to Parser](https://github.com/wooorm/remark/blob/bb13322e07bf9173724eabbce229770387c12c86/test/mentions.js#L92-L98) (or vice versa, really).

```js
function attacher(remark) {
    var proto = remark.Parser.prototype;
    var methods = proto.inlineMethods;

    /*
     * Add a tokenizer to the `Parser`.
     */

    proto.inlineTokenizers.mention = mention;

    methods.splice(methods.indexOf('inlineText'), 0, 'mention');
}
```

It would be nice if a compiler for mentions could hook itself this way:

```js
function attacher(remark) {
    var proto = remark.Compiler.prototype;

    /*
     * Add a visitor to the `Compiler`.
     */

    proto.visitors.mention = mention;
}
```

I'd love that kind of symmetry in place :smile:

**Comment by wooorm** [comment]
:+1:

One final thing: any reason to *not* opt for the name `visitors`? Alternatives: `compilers`, `visitor`, ...

@ben-eb, with your PostCSS experience, could you shed some light on your opinion regarding this (if you don’t really care that’s also fine :wink:)?

**Comment by ben-eb** [comment]
We use much the same as the current system, although the node selection is a lot more limited; `atrule`, `rule`, `decl`, `comment`, `root`.

**Comment by wooorm** [comment]
@ben-eb I can imagine. Is there support (or reason for) for custom nodes?

**Comment by ben-eb** [comment]
Not really, mostly everything already fits into the CSS grammar as defined. We haven't needed to add any other node types to support SCSS grammar, for instance.

**Comment by wooorm** [comment]
Okay, thanks!
